### DISCUSSION TITLE
profile="alpha_j"

### DISCUSSION THREAD

**Discussion by MahtabYazdanian** [vast-lib/tinyVAST#53] [Q&A]
Hi Jim,

What does profile = "alpha_j" argument determine in tinyVAStcontrol?

I want to get deviance explained, but it is only possible if control$profile is not specified.

**Comment by James-Thorson-NOAA** [comment]
`profile` pushes some fixed effects into the inner-loop of the optimizer, while excluding them from the Hessian used to compute the Laplace approximation.  See Chap-2 of our book if you wanna see more context.  An RTMB demo is here:

```R
library(RTMB)
X = runif(100)
Y = 1 + 0.5 * X + rnorm(100)

nll = function(p){
  -1*sum(dnorm(Y, p$b0 + p$b1*X, exp(p$logsd), log=TRUE))
}
p = list(b0=0, b1=0, logsd=0)
nll(p)
obj = MakeADFun(func=nll, parameters=p, profile=c("b0","b1"))
opt = do.call(optim, obj)
```

I just pushed an edit to the tinyVAST@dev branch that allows the proportion conditional deviance explained (PCDE) to be calculated even when profiling across all fixed effects (presumably because you're fitting a Poisson count-data model).  Could you update and see if that fixes it for you, or just source the dev-branch `deviance_explained` function and use it with your existing version?   See demo here:

```R
library(tinyVAST)
Z = rnorm(10)
group = rep(1:10,each=10)
Y = rpois(100, lambda=exp(0 + Z[group]) )

myfit = tinyVAST(
  formula = Y ~ 1 + s(group,bs="re"),
  data = data.frame(Y=Y, group=factor(group)),
  family = poisson(link="log"),
  control = tinyVASTcontrol(profile="alpha_j")
)
source( file.path(R'(C:\Users\James.Thorson\Desktop\Git\tinyVAST\R)',"utility.R") )
deviance_explained(myfit)
```

Also note that in a JSDM (as you are perhaps using) it may make sense to compute the PCDE relative to a model with a separate intercept for each species.  This is more akin to the PCDE relative to a stacked GLM with intercepts for each individual GLM.  see `?deviance_explained` for the arguments allowing you to custom-specify the null formula.  

I'm closing for now but happy to keep discussing here.

**Reply by MahtabYazdanian** [reply]
Thank you Jim for the demo. In the tinyVAST demo, is it so that group-level random effects are profiled?

I sourced the dev-branch, but still ran into this error when calling deviance_explained():

Error in nlminb(start = opt$par, objective = obj$fn, gradient = obj$gr,  : 
  'd' must be a nonempty numeric vector

**Comment by James-Thorson-NOAA** [comment]
Thanks for flagging it!  Do you mind emailing me another minimal example?  It was working on the previous example when I tested it.

**Reply by MahtabYazdanian** [reply]
I reinstalled the dev branch and now it works. It did not work when I sourced.

Does profiling, or more specifically profiling alpha-j have more benefit besides speeding up model convergence? In my case, models with and without profiling have very similar predictions, so I was wondering how to choose between them.

**Comment by James-Thorson-NOAA** [comment]
Yeah I think they should be identical except for speed. How are they
different?

On Thursday, October 24, 2024, Matoosh ***@***.***> wrote:

> I reinstalled the dev branch and now it works. It did not work when I
> sourced.
>
> Does profiling, or more specifically profiling alpha-j have more benefit
> besides speeding up model convergence? In my case, models with and without
> profiling have very similar predictions, so I was wondering how to choose
> between them.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vast-lib/tinyVAST/discussions/53#discussioncomment-11039443>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AL62VMSU6XSWTCO7FFD25KTZ5DDDNAVCNFSM6AAAAABQFI45MCVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMBTHE2DIMY>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>


-- 
James Thorson (he/him)

Alaska Fisheries Science Center, NMFS
Affiliate Faculty, University of Washington

**Reply by MahtabYazdanian** [reply]
Parameter estimates are different. See the summary of two models attached.
[Summary-NotProfiled.txt](https://github.com/user-attachments/files/17522380/Summary-NotProfiled.txt)
[Summary-Profiled.txt](https://github.com/user-attachments/files/17522382/Summary-Profiled.txt)

**Comment by James-Thorson-NOAA** [comment]
Yeah, interesting.  Could you check whether the estimates remain different when getting really tight in the MLE, i.e., using `tinyVASTcontrol(newton_loops = 2)`?  

If they remain different, I could then do some digging.

**Comment by MahtabYazdanian** [comment]
I checked, and I'd say the differences got much smaller but still a bit different. The predictions are still highly correlated (r<0.99)

[MLE_summary_NotProfiled.txt](https://github.com/user-attachments/files/17598731/MLE_summary_NotProfiled.txt)
[MLE_summary_Profiled.txt](https://github.com/user-attachments/files/17598732/MLE_summary_Profiled.txt)
).

**Comment by James-Thorson-NOAA** [comment]
And do you still have the actual `logLik(myfit)` value for each run too? 

It certainly looks like those are tightly converged, and have a larger difference in MLE than I expected.  The `profile` feature is relatively new to me, so I'll investigate and post a question to the TMB listserve today.

In the meantime, I'd either use the one without profile, or use the profile argument as a first pass, then pass the MLE back in as starting value to one without profile.

Jim

**Comment by James-Thorson-NOAA** [comment]
OK, the response I got was that there's more theory [here](http://kaskr.github.io/adcomp/_book/Appendix.html#profiling-the-inner-problem).  I'll have to think about how to proceed in terms of documentation, but my advise is to proceed without the profile argument for now, and only use it when the differences are trivial.

**Reply by MahtabYazdanian** [reply]
Ok, I will go ahead as you suggested. the logLik() for the model not profiled is -7698.791 (df=118) and the other one is -7695.22 (df=118).
